backend/src/packages/chaiNNer_ncnn/ncnn/processing/upscale_image.py

```diff
@@ -70,18 +70,33 @@ def upscale_impl(
     net = get_ncnn_net(model, exec_options)
     # Try/except block to catch errors
     try:
-        if use_gpu:
-            vkdev = ncnn.get_gpu_device(exec_options.ncnn_gpu_index)
 
-            def estimate_gpu():
+        def estimate():
+            heap_budget = exec_options.ncnn_budget_limit
+            model_size_estimate = model.model.bin_length
+
+            if use_gpu:
                 if is_mac:
                     # the actual estimate frequently crashes on mac, so we just use 256
                     return MaxTileSize(256)
 
-                heap_budget = vkdev.get_heap_budget() * 1024 * 1024 * 0.8
-                return MaxTileSize(
-                    estimate_tile_size(heap_budget, model.model.bin_length, img, 4)
+                heap_budget = min(
+                    heap_budget, vkdev.get_heap_budget() * 1024 * 1024 * 0.8
                 )
+            else:
+                # Empirically determined
```

Comment:
Since you have so much RAM, I'm wondering if you'd be willing to run some tests. I've done some testing on NCNN VRAM estimation in the past (though only with ESRGAN), and I can already tell you that this isn't really correct, and neither is what we already have in place. I've only got 8GB VRAM, so I could never extend my tests far enough to gather fully complete data.

What I can say is that:
1) Model size is not strongly correlative with VRAM usage. This is because a model can be made larger simply by performing more convolutions, which does not matter for VRAM usage because they are being done in sequence. The only thing it definitely correlates with is how much VRAM it takes to store the model itself.
2) Individual weight sizes have a correlation with VRAM usage when running a model.
3) Scale needs to be accounted for, which our estimation does not currently do. This estimation is based around a 4x scale, but an 8x model will blow past it.

I abandoned this back in the day because there seemed to be further factors I couldn't account for with the data I had, but maybe we can finally figure it out. Unfortunately, I seem to have deleted the set of different scale/nf/nb ESRGAN models I generated for these tests. If I can remember how I generated them all, I could send them to you.

Reply:
@theflyingzamboni I would definitely be interested. Can we split this out into its own issue so that it doesn't get lost once this PR is merged?